### PR TITLE
Completely remove Collection/ReadonlyCollection types

### PR DESCRIPTION
**Pull request by jakebailey**
Split out of #51921 to constrain it to more valuable changes.

### CODE REVIEW

**Review by sandersn — APPROVED** [review]

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -675,7 +674,7 @@ export function forEachEntry<K, V, U>(map: ReadonlyMap<K, V>, callback: (value:
  *
  * @internal
  */
-export function forEachKey<K, T>(map: ReadonlyCollection<K>, callback: (key: K) => T | undefined): T | undefined {
+export function forEachKey<K, T>(map: ReadonlyMap<K, unknown> | ReadonlySet<K>, callback: (key: K) => T | undefined): T | undefined {
```
Comment (jakebailey) [inline src/compiler/utilities.ts:677]: This one was funky; swap the argument types and inference breaks. Need to retest to see if my inference fixups fixed this, though.

Comment (jakebailey) [reply src/compiler/utilities.ts:677]: Nah, still screwy. I think I'll file a bug.

Comment (jakebailey) [reply src/compiler/utilities.ts:677]: #52313

**Comment by sandersn** [comment]
@jakebailey Same here; let's merge this for 5.1 if that last comment isn't a concern.

**Comment by jakebailey** [comment]
I think it's a concern, in that if it's order dependent now, it might break in the future if the order internally changes somehow. So, I'd like the bug to be looked at first.

**Comment by jakebailey** [comment]
Closing this since there's no way I'd want to merge this before #52313 is fixed, which itself looks to be a challenging bug.

**Pull request closed without merging**